### PR TITLE
Install and use Homebrew's `ca-certificates` on macOS <= 10.15.5

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

The final step. Must be merged after #12166.

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]
😭 👍🏻

**Comment by Bo98** [comment]
Following discussion https://github.com/Homebrew/homebrew-core/pull/86291#issuecomment-932841337, I've changed this PR so that we only install one formula: `ca-certificates` (arriving soon to homebrew-core) and then set `SSL_CERT_FILE`. This allows system curl to still be used, which greatly simplifies the situation with homebrew-test-bot.

**Comment by Bo98** [comment]
One caveat with this new approach is that the SSL bug _is_ still present so this issue can resurface in the future.

If this happens, it can be fixed by `brew postinstall ca-certificates` or we can revision bump the formula to force this to happen.

File: Library/Homebrew/cmd/update.sh [hunk 1]
```diff
@@ -375,6 +375,13 @@ user account:
 EOS
   fi
 
+  # we may want to use Homebrew CA certificates
+  if [[ -n "${HOMEBREW_FORCE_BREWED_CA_CERTIFICATES}" && ! -f "${HOMEBREW_PREFIX}/etc/ca-certificates/cert.pem" ]]
+  then
+    # we cannot install Homebrew CA certificates if homebrew/core is unavailable.
+    ([[ -d "${HOMEBREW_LIBRARY}/Taps/homebrew/homebrew-core" ]] && brew install ca-certificates) || true
```
Comment (carlocab) [inline Library/Homebrew/cmd/update.sh:382]:
```suggestion
    ([[ -d "${HOMEBREW_LIBRARY}/Taps/homebrew/homebrew-core" || -n "$HOMEBREW_INSTALL_FROM_API" ]] && brew install ca-certificates) || true
```
Maybe?

**Review by carlocab — APPROVED** [review]

File: Library/Homebrew/cmd/update.sh (hunk 1 above)
Comment (Bo98) [reply Library/Homebrew/cmd/update.sh:382]: Looks like that makes sense, though the curl and git stuff don't have this yet. cc @Rylan12

**Comment by Bo98** [comment]
Hmm SecureTransport on 10.10 still doesn't seem to work even with this. Might end up with a hybrid here where some really old versions need a full brewed curl but Mojave etc can live with just ca-certificates - will investigate more tomorrow.

**Review by Rylan12 — APPROVED** [review]
Looks good to me. At some point, it will be great to check that the `brew install` calls here work with `HOMEBREW_INSTALL_FROM_API` but that isn't a priority at the moment.

**Comment by carlocab** [comment]
> Hmm SecureTransport on 10.10 still doesn't seem to work even with this. Might end up with a hybrid here where some really old versions need a full brewed curl but Mojave etc can live with just ca-certificates - will investigate more tomorrow.

Given @EricFromCanada's tests this seems to work on Sierra or newer. I suggest we ship this as soon as we can to fix users who are on those systems. We can always work on a fix for older OSs after.

**Comment by Rylan12** [comment]
> I suggest we ship this as soon as we can

We need to wait for https://github.com/Homebrew/homebrew-core/pull/86304 gets through, first

Also just a reminder that this will need a new `brew` tag

**Comment by carlocab** [comment]
Also, I wonder if this will allow us to get rid of `using: :homebrew_curl`.

**Comment by Bo98** [comment]
> Also, I wonder if this will allow us to get rid of `using: :homebrew_curl`.

I wasn't around when this was added. Why was this necessary? Which platforms did this affect?

**Comment by carlocab** [comment]
Cert errors, I think. @iMichka should know the details.

**Comment by Bo98** [comment]
Ok xiph seems to require TLS 1.3, which macOS LibreSSL does not support.

**Comment by Bo98** [comment]
Ok so I've added an extra < Sierra check, which will install a full brewed curl. 10.12-10.15.6 will have the lightweight `ca-certificates` route.

**Comment by Rylan12** [comment]
> Ok so I've added an extra < Sierra check, which will install a full brewed curl. 10.12-10.15.5 will have the lightweight `ca-certificates` route.

Sorry if I just missed this, but why won't the `ca-certificates` route work on < Sierra?

**Comment by Bo98** [comment]
It seems there's some Secure Transport bug or something. Tests on 10.10 reveal that ftp.gnu.org just returns "not trusted" even when passed the ISRG Root X1 alone. Hard to get any meaningful information from Secure Transport - it has poor debugging support.

My _guess_ is it it doesn't handle alernative chains properly.

Eric tested all macOS versions and found it works on 10.12 and later.

**Comment by Bo98** [comment]
Basically we're seeing https://github.com/curl/curl/issues/976 on < Sierra.

**Review by Rylan12 — APPROVED** [review]

**Comment by Bo98** [comment]
> Must be merged after #12166.

I hate chickens, and I hate eggs.

This can't happen because it needs the `nghttp2` PR, but the `nghttp2` PR fails because it needs to revision bump formulae outside the curl dep tree and they use HTTPS, which is broken and needs this PR, and the brewed curl test-bot workaround also doesn't work because the `nghttp2` PR revision bumps brewed curl itself.

Lost? I don't blame you.

Luckily `nghttp2` and its remaining deps all use GitHub so it should be ok.

**Comment by carlocab** [comment]
Here's the first successful Homebrew/core CI job from this change: https://github.com/Homebrew/homebrew-core/pull/86421/checks?check_run_id=3788099599

I needed Homebrew/homebrew-test-bot@4fd658e8456f74da1131296ebd8ddb421a9327f4 before it worked though. Not sure why.

**Comment by Bo98** [comment]
> I needed [Homebrew/homebrew-test-bot@4fd658e](https://github.com/Homebrew/homebrew-test-bot/commit/4fd658e8456f74da1131296ebd8ddb421a9327f4) before it worked though. Not sure why.

This'll be because we never export `HOMEBREW_FORCE_BREWED_CA_CERTIFICATES`, so it can't be checked in Ruby.

Either we start exporting it or we instead check `DevelopmentTools.ca_file_handles_most_https_certificates?`.

(Same scenario with `curl`, even though we don't really need it for any of our CI.)

File: Library/Homebrew/cmd/update.sh [hunk 2]
```diff
@@ -375,11 +375,21 @@ user account:
 EOS
   fi
 
+  # we may want to use Homebrew CA certificates
+  if [[ -n "${HOMEBREW_FORCE_BREWED_CA_CERTIFICATES}" && ! -f "${HOMEBREW_PREFIX}/etc/ca-certificates/cert.pem" ]]
+  then
+    # we cannot install Homebrew CA certificates if homebrew/core is unavailable.
+    if [[ -d "${HOMEBREW_LIBRARY}/Taps/homebrew/homebrew-core" || -n "${HOMEBREW_INSTALL_FROM_API}" ]]
+    then
+      brew install ca-certificates || true
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/update.sh:384]: the `|| true` seems undesirable here, we want to handle if this fails (with at least a message), no?

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks for this!

File: Library/Homebrew/cmd/update.sh (hunk 2 above)
Comment (Bo98) [reply Library/Homebrew/cmd/update.sh:384]: I had a reason in my head before but so much has changed that I can't remember it and it probably no longer applies.

I'm not sure what the message would be though. The solution for brewed curl and git is to install it yourself and put it in the PATH. This doesn't apply to CA certificates.

Comment (MikeMcQuaid) [reply Library/Homebrew/cmd/update.sh:384]: I think it's worth just failing in this case, then.

**Comment by carlocab** [comment]
> Either we start exporting it or we instead check `DevelopmentTools.ca_file_handles_most_https_certificates?`.

Homebrew/homebrew-test-bot#674

**Pull request merged**